bee-protocol/src/workers/peer/packet_handler.rs

```diff
@@ -0,0 +1,333 @@
+
+use bee_network::Multiaddr;
+
+use futures::{
+    channel::oneshot,
+    future::{self, FutureExt},
+    stream::StreamExt,
+};
+use log::trace;
+use tokio::select;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+type EventRecv = UnboundedReceiverStream<Vec<u8>>;
+type ShutdownRecv = future::Fuse<oneshot::Receiver<()>>;
+
+/// The read state of the packet handler.
+///
+/// This type is used by `PacketHandler` to decide what should be read next when handling an
+/// event.
+enum ReadState {
+    /// `PacketHandler` should read a header.
+    Header,
+    /// `PacketHandler` should read a payload based on a header.
+    Payload(HeaderPacket),
+}
+
+/// A packet handler.
+///
+/// It takes care of processing events into packets that can be processed by the workers.
+pub(super) struct PacketHandler {
+    events: EventHandler,
+    // FIXME: see if we can implement `Stream` for the `PacketHandler` and use the
+    // `ShutdownStream` type instead.
+    shutdown: ShutdownRecv,
+    state: ReadState,
+    /// The address of the peer. This field is only here for logging purposes.
+    address: Multiaddr,
+}
+
+impl PacketHandler {
+    /// Create a new packet handler from an event receiver, a shutdown receiver and the peer's
+    /// address.
+    pub(super) fn new(receiver: EventRecv, shutdown: ShutdownRecv, address: Multiaddr) -> Self {
+        Self {
+            events: EventHandler::new(receiver),
+            shutdown,
+            // The handler should read a header first.
+            state: ReadState::Header,
+            address,
+        }
+    }
+    /// Fetch the header and payload of a packet.
+    ///
+    /// This method only returns `None` if a shutdown signal is received.
+    pub(super) async fn fetch_packet(&mut self) -> Option<(HeaderPacket, &[u8])> {
+        // loop until we can return the header and payload
+        loop {
+            match &self.state {
+                // Read a header.
+                ReadState::Header => {
+                    // We need `HEADER_SIZE` bytes to read a header.
+                    let bytes = self
+                        .events
+                        .fetch_bytes_or_shutdown(&mut self.shutdown, HEADER_SIZE)
+                        .await?;
+                    trace!("[{}] Reading Header...", self.address);
+                    let header = HeaderPacket::from_bytes(bytes);
+                    // Now we are ready to read a payload.
+                    self.state = ReadState::Payload(header);
+                }
+                // Read a payload.
+                ReadState::Payload(header) => {
+                    // We read the quantity of bytes stated by the header.
+                    let bytes = self
+                        .events
+                        .fetch_bytes_or_shutdown(&mut self.shutdown, header.packet_length.into())
+                        .await?;
+                    // FIXME: Avoid this clone
```

Comment:
Oh Christian with the longer telomeres what did you have in mind?

Reply:
@thibault-martinez what would you say about using: https://docs.rs/take_mut/0.2.2/take_mut/ so we can swap the `self.state` value without using unsafe nor cloning?